New plugin to read RAID events

mfijalko opened the pull request:
This plugin, named **mdevents**,  is responsible for gathering the events from RAID arrays that were written to syslog by mdadm utility (which is a user-space software for managing the RAIDs). Then, based on configuration provided by user, plugin will decide whether to send the collectd notification or not. 

Mdevents needs the syslog and mdadm to be present on a platform that collectd is launched.

Based on the naming of existing dpdk plugins (dpdkstat for metrics, dpdkevents for notifications) and the fact that there is already plugin called **md** for gathering metrics from RAIDs, giving this plugin name **mdevents** felt like the best option.

A slight change to ignorelist API was introduced - it is now possible to retrieve the count of nodes in array and the current **ignore** setting for given array.

ChangeLog: mdevents plugin: Generate notifications about events in RAID arrays.

[edit @kkepka:] added changelog

kwiatrox commented on configure.ac:
```diff
@@ -6328,6 +6335,7 @@ if test "x$ac_system" = "xLinux"; then
   plugin_fhcount="yes"
   plugin_fscache="yes"
   plugin_hugepages="yes"
+  plugin_mdevents="yes"
```
This variable is not used. Check for `enable_mdevents == "yes"` and set `plugin_mdevents` basing on that condition.

kwiatrox commented on configure.ac:
```diff
@@ -6695,6 +6703,7 @@ AC_PLUGIN([hddtemp],             [yes],                     [Query hddtempd])
 AC_PLUGIN([hugepages],           [$plugin_hugepages],       [Hugepages statistics])
 AC_PLUGIN([intel_pmu],           [$with_libjevents],        [Intel performance monitor plugin])
 AC_PLUGIN([intel_rdt],           [$with_libpqos],           [Intel RDT monitor plugin])
+AC_PLUGIN([mdevents],            [$enable_mdevents],        [Events from md (Linux Software RAID) devices])
```
Check for `plugin_mdevents` instead of `enable_mdevents`

kwiatrox commented on configure.ac:
```diff
@@ -7116,6 +7125,7 @@ AC_MSG_RESULT([    hddtemp . . . . . . . $enable_hddtemp])
 AC_MSG_RESULT([    hugepages . . . . . . $enable_hugepages])
 AC_MSG_RESULT([    intel_pmu . . . . . . $enable_intel_pmu])
 AC_MSG_RESULT([    intel_rdt . . . . . . $enable_intel_rdt])
+AC_MSG_RESULT([    mdevents  . . . . . . $enable_mdevents])
```
Check for `plugin_mdevents` instead of `enable_mdevents`

kwiatrox commented on src/mdevents.c:
```diff
@@ -0,0 +1,428 @@
+#define MD_EVENTS_PLUGIN "mdevents"
+#define DAEMON_NAME "mdadm"
+
+#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof(arr[0]))
```
In collectd already exists predefined macro: STATIC_ARRAY_SIZE

kwiatrox commented on src/mdevents.c:
```diff
@@ -0,0 +1,428 @@
+  // monitor events only from point of collectd start
+  fseek(syslog_file, 0, SEEK_END);
+
+  if (md_events_compile_regex(&regex, regex_pattern))
```
fclose(syslog_file)

kwiatrox commented on src/mdevents.c:
```diff
@@ -0,0 +1,428 @@
+  md_events_copy_match(event.md_device, to_match, matches[MD_DEVICE]);
+
+  // this one is not present in every event, regex API sets indexes to -1
+  // if the match wasn't found
```
Maybe return 0 if match is not found in the current line.

kwiatrox commented on src/mdevents.c:
```diff
@@ -0,0 +1,428 @@
+  }
+
+  // monitor events only from point of collectd start
+  fseek(syslog_file, 0, SEEK_END);
```
Optionally check for `fseek` return value.

mfijalko commented:
> What do you think about implementation, which will not require syslog in the chain?
> Unfortunately, I found no such solution. Maybe we can ask `mdadm` developers to start implementing the library for such purposes?

Hi! First of all, sorry for that prolonged silence. Thanks for taking a look at this. I'm going to post a patch that addresses your comments.

Answering your question - from what I see, `mdadm` doesn't look like it is "alive", their github is silent for more or less half a year. But you know, we can give it a shot and post a feature request.

On the other hand, does having a syslog dependency is that painful? 

Thanks again and looking forward for your response!

rpv-tomsk commented:
This PR is something like or related to #3045

rpv-tomsk commented:
>On the other hand, does having a syslog dependency is that painful?

The longer chain - lower  reliability.

kkepka commented:
> This PR is something similar to or related to #3045

You are right @rpv-tomsk, functionally both plugins are watching files for particular messages to generate notification about.
The difference is that `mdevents` is to be just enabled and ready to go, while in `logparser` case user needs to provide definition of what want to monitor (via proper configuration). So it's trade of easy-to-use vs flexibility and being error prone on configuration side.

dago commented:
@mfijalko There is a failing test, would you mind having a look?

kkepka commented:
Red Hat check have more verbose UT logs.
From there we can see, that UT fails for init unit test when there is no `/var/log/syslog` or `/var/log/messages` available on the system, what is the case in ci env, so i'd say it's more unitests implementation / test env issue, rather then plugin itself.
I don't want to create any debt here, but as ci environment doesn't have dependency for this unit test case, would it make sense to remove it?

sunkuranganath commented:
Looks like all the changes requested has been resolved along with any of the UT/CI issues. Could we remove "changes requested"?

dago commented:
@sunkuranganath At least I cannot. @rpv-tomsk Can you resolve the changerequest?
